xla/codegen/emitters/transforms/lower_tensors.cc

```diff
@@ -1405,9 +1405,39 @@ class LowerTensorsPass : public impl::LowerTensorsPassBase<LowerTensorsPass> {
         signalPassFailure();
       }
     });
+    
+    // For AMDGPU, ensure allocas use address space 5 (private)
+    if (device_spec_.IsAmdGpu()) {
+      EnsureAMDGPUAllocasUseAS5();
+    }
   }
 
  private:
+  // Ensure AMDGPU allocas use address space 5 (private).
+  // AMDGPU requires allocas in AS5, but MLIR lowering creates them in AS0.
+  void EnsureAMDGPUAllocasUseAS5() {
```

Comment:
Please don't  duplicate this function. Instead, create two new files, e.g. `lowering_utils.h`/`.cc` and add this functionality there.

Sorry I didn't provide this comment earlier.